### ISSUE TITLE
Having trouble getting started

### ISSUE THREAD

**Issue by zimmertr** [sergelogvinov/proxmox-csi-plugin#93]
Hello, first of all I'd like to give you an enormous THANK YOU for putting this together! I've been using NFS for Kubernetes and Proxmox for years and it hasn't been great. This looks like it will solve all my problems. However, I'm having some trouble getting it up and running and I was hoping you might be able to help me if I abused your issue board a bit? 😅 

The CSI plugin is installed and running:

```bash
$> kubectl get all -n csi-proxmox
NAME                                                 READY   STATUS    RESTARTS   AGE
pod/proxmox-csi-plugin-controller-6d5dd8856d-4f9bp   5/5     Running   0          2m56s

NAME                                     DESIRED   CURRENT   READY   UP-TO-DATE   AVAILABLE   NODE SELECTOR                                       AGE
daemonset.apps/proxmox-csi-plugin-node   0         0         0       0            0           node.cloudprovider.kubernetes.io/platform=nocloud   2m56s

NAME                                            READY   UP-TO-DATE   AVAILABLE   AGE
deployment.apps/proxmox-csi-plugin-controller   1/1     1            1           2m56s

NAME                                                       DESIRED   CURRENT   READY   AGE
replicaset.apps/proxmox-csi-plugin-controller-6d5dd8856d   1         1         1       2m56s
```

Using the following `config.yaml`:

```yaml
clusters:
  - url: https://earth.sol.milkyway:8006/api2/json
    insecure: true
    token_id: "kubernetes-csi@pve!csi"
    token_secret: "REDACTED"
    region: cluster-1
```

I've deployed a `StorageClass` called `local` that is referencing a `Storage` on Proxmox called `local` that is a _Directory_ type that allows _all content_. (I also tried the `FlashPool` ZFS pool that allows Disk Image and Container unsuccessfully too)

![Screenshot 2023-10-15 at 4 55 25 PM](https://github.com/sergelogvinov/proxmox-csi-plugin/assets/4356289/be1879f8-e5a5-4a87-8d9c-4c642feb1e7b)

```yaml
$> kubectl get sc local -oyaml
allowVolumeExpansion: true
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  name: local
parameters:
  storage: local
provisioner: csi.proxmox.sinextra.dev
reclaimPolicy: Delete
volumeBindingMode: Immediate
```

However, when I deploy the test pods (modified yaml to use my StorageClass, of course), they are stuck pending. If I describe the created PVCs, I see:

```bash
Events:
  Type     Reason                Age                  From                                                                                                          Message
  ----     ------                ----                 ----                                                                                                          -------
  Normal   Provisioning          55s (x9 over 5m10s)  csi.proxmox.sinextra.dev_proxmox-csi-plugin-controller-6d5dd8856d-4f9bp_4f17b04e-fbc9-42b4-a3cc-eb455116a355  External provisioner is provisioning volume for claim "default/test-pvc"
  Warning  ProvisioningFailed    55s (x9 over 5m10s)  csi.proxmox.sinextra.dev_proxmox-csi-plugin-controller-6d5dd8856d-4f9bp_4f17b04e-fbc9-42b4-a3cc-eb455116a355  **failed to provision volume with StorageClass "local": error generating accessibility requirements: no available topology found**
  Normal   ExternalProvisioning  3s (x22 over 5m10s)  persistentvolume-controller                                                                                   Waiting for a volume to be created either by the external provisioner 'csi.proxmox.sinextra.dev' or manually by the system administrator. If volume creation is delayed, please verify that the provisioner is running and correctly registered.
```

So the error is `no available topology found`.  I have labeled my nodes to all include both: `topology.kubernetes.io/region=cluster-1` and `topology.kubernetes.io/zone=earth` 

```bash
$> kubectl get nodes --show-labels
NAME         STATUS   ROLES           AGE   VERSION   LABELS
k8s-cp-1     Ready    control-plane   8h    v1.28.2   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/os=linux,kubernetes.io/arch=amd64,kubernetes.io/hostname=k8s-cp-1,kubernetes.io/os=linux,node-role.kubernetes.io/control-plane=,topology.kubernetes.io/region=cluster-1,topology.kubernetes.io/zone=earth
k8s-cp-2     Ready    control-plane   8h    v1.28.2   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/os=linux,kubernetes.io/arch=amd64,kubernetes.io/hostname=k8s-cp-2,kubernetes.io/os=linux,node-role.kubernetes.io/control-plane=,topology.kubernetes.io/region=cluster-1,topology.kubernetes.io/zone=earth
k8s-cp-3     Ready    control-plane   8h    v1.28.2   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/os=linux,kubernetes.io/arch=amd64,kubernetes.io/hostname=k8s-cp-3,kubernetes.io/os=linux,node-role.kubernetes.io/control-plane=,topology.kubernetes.io/region=cluster-1,topology.kubernetes.io/zone=earth
k8s-node-1   Ready    <none>          8h    v1.28.2   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/os=linux,kubernetes.io/arch=amd64,kubernetes.io/hostname=k8s-node-1,kubernetes.io/os=linux,topology.kubernetes.io/region=cluster-1,topology.kubernetes.io/zone=earth
k8s-node-2   Ready    <none>          8h    v1.28.2   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/os=linux,kubernetes.io/arch=amd64,kubernetes.io/hostname=k8s-node-2,kubernetes.io/os=linux,topology.kubernetes.io/region=cluster-1,topology.kubernetes.io/zone=earth
k8s-node-3   Ready    <none>          8h    v1.28.2   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/os=linux,kubernetes.io/arch=amd64,kubernetes.io/hostname=k8s-node-3,kubernetes.io/os=linux,topology.kubernetes.io/region=cluster-1,topology.kubernetes.io/zone=earth
 ```
 
Which reflect the name of my Proxmox node (earth) and the single-node cluster I created (cluster-1).

![Screenshot 2023-10-15 at 4 58 03 PM](https://github.com/sergelogvinov/proxmox-csi-plugin/assets/4356289/05b2b752-e310-46b4-ac2a-c4b6808a7c32)

In the README I noticed the `Usage` section mentions getting several CRDs. Presumably for troubleshooting? If I do this, I do see some bad results. For example:

CSI Driver looks good....

```bash
$> kubectl get CSIDriver
NAME                       ATTACHREQUIRED   PODINFOONMOUNT   STORAGECAPACITY   TOKENREQUESTS   REQUIRESREPUBLISH   MODES        AGE
csi.proxmox.sinextra.dev   true             true             true              <unset>         false               Persistent   8m17
```

CSIStorageCapacties is null?

```bash
$> kubectl get csistoragecapacities.storage.k8s.io -A
No resources found
```

CSINode has a null spec.drivers

```bash
$> kubectl get csinode k8s-node-1 -oyaml
apiVersion: storage.k8s.io/v1
kind: CSINode
metadata:
  annotations:
    storage.alpha.kubernetes.io/migrated-plugins: kubernetes.io/aws-ebs,kubernetes.io/azure-disk,kubernetes.io/azure-file,kubernetes.io/cinder,kubernetes.io/gce-pd,kubernetes.io/vsphere-volume
  creationTimestamp: "2023-10-15T15:03:25Z"
  name: k8s-node-1
  ownerReferences:
  - apiVersion: v1
    kind: Node
    name: k8s-node-1
    uid: b59fbe9a-d2da-4a2a-9953-5e8efd42ec39
  resourceVersion: "355"
  uid: b4b2fd45-5e2b-4095-846b-6b2f072aa9af
spec:
  drivers: null
```

<hr>

And that's about as far as I've gotten after several hours of troubleshooting and reading Go code lol. I'm stumped. Hoping you might have some advice or ideas? Is there no support for single-node clusters maybe? A bad config somewhere?

**Comment by sergelogvinov** [comment]
Hello `csistoragecapacities.storage.k8s.io` is empty, probably your proxmox-token does not have permission.

May be these commands help, or use root token (temporary):
```shell
pveum aclmod / -user 'kubernetes@pve' -role CSI
pveum aclmod / -tokens 'kubernetes@pve!csi' -role CSI
```

**Comment by zimmertr** [comment]
Hm, I just ran the `pveum` commands listed in your README:

```bash
pveum role add CSI -privs "VM.Audit VM.Config.Disk Datastore.Allocate Datastore.AllocateSpace Datastore.Audit"
pveum user add kubernetes-csi@pve
pveum aclmod / -user kubernetes-csi@pve -role CSI
pveum user token add kubernetes-csi@pve csi -privsep 0
```

So I didn't run `pveum aclmod / -tokens 'kubernetes-csi@pve!csi' -role CSI`. However, I created a new token for the user and ran those commands again and haven't gotten any further. Here is an end-to-end screenshot of my creating a new token, running those commands, echoing the base64 encoded config to show the token, and getting the resources again:

![Screenshot 2023-10-16 at 1 56 36 PM](https://github.com/sergelogvinov/proxmox-csi-plugin/assets/4356289/679723bb-2362-48dd-82c4-a2ffc3ac4e84)

What is supposed to be creating the `csistoragecapacities.storage.k8s.io` resource(s)? Is there any chance I can grab logs or Kubernetes events from it to dig a little deeper? 

Here are the logs from the five containers in the `proxmox-csi-plugin-controller` pod. I don't see any obvious errors but perhaps I don't know what to look for? 

* csi-attacher: [here](https://gist.github.com/zimmertr/c82013f1afd69656e4a2c5f3a2693fde)
* csi-provisioner: [here](https://gist.github.com/zimmertr/66d21f26a057ba6d6c7fc2ccd57f83eb)
* csi-resizer: [here](https://gist.github.com/zimmertr/e650e8451299f5d77afa05edd4082e33)
* liveness-probe: [here](https://gist.github.com/zimmertr/fb116d389f864377ff0f6ba0759a82c9)
* proxmox-csi-plugin-controller: [here](https://gist.github.com/zimmertr/e07507ea9e82071fb42a0c89df0b55f3)

Here is also a link to a full `describe` on the pod itself: [here](https://gist.github.com/zimmertr/5862aed49fae6d5eeeceb731ab44645f)

**Comment by zimmertr** [comment]
I also tried using the root user as well. I assume I don't need to give it any permissions with `aclmod`? Let me know if I'm wrong about that and I can try again. 

![image](https://github.com/sergelogvinov/proxmox-csi-plugin/assets/4356289/7b0f7316-9d23-45a0-a3dd-a02e3b6507ad)

**Comment by zimmertr** [comment]
(Also yes I know I'm leaking secrets for my Proxmox server but it's not exposed to the internet and I also removed the tokens before posting here 😉 )

**Comment by sergelogvinov** [comment]
Oh, now i see, csi-plugin-node has size "0" 
Can you label the node `node.cloudprovider.kubernetes.io/platform: nocloud`
this label sets TalosCCM automatically....

**Comment by LeoShivas** [comment]
Hi,

I've encountered the same issues as yours till right now.

I've managed to made it work by doing the following :
### Delete all ###
Delete all the proxmox-csi-plugin and proxmox-ccm related :
storage class
namespace
secrets
and all the stuff you've created for the proxmox-csi/ccm

### Create the Proxmox users ###
As the documentation says :
```
pveum user add kubernetes-csi@pve
pveum role add CSI -privs "VM.Audit VM.Config.Disk Datastore.Allocate Datastore.AllocateSpace Datastore.Audit"
pveum aclmod / -user kubernetes-csi@pve -role CSI
pveum user token add kubernetes-csi@pve csi -privsep 0
pveum role add CCM -privs "VM.Audit"
pveum user add kubernetes@pve
pveum aclmod / -user kubernetes@pve -role CCM
pveum user token add kubernetes@pve ccm -privsep 0
```
Be aware, don't invert the CCM and the CSI account. They do not have the same rights.

### Create the namespace ###
Create the namespace with the following file :
```
# csi-ns.yml
apiVersion: v1
kind: Namespace
metadata:
  name: csi-proxmox
  labels:
    pod-security.kubernetes.io/enforce: privileged
    pod-security.kubernetes.io/audit: baseline
    pod-security.kubernetes.io/warn: baseline
    app.kubernetes.io/managed-by: Helm
  annotations:
    meta.helm.sh/release-name: proxmox-csi-plugin
    meta.helm.sh/release-namespace: csi-proxmox
```

`kubectl apply -f csi-ns.yml`

### Create the CCM secret ###
Follow the doc (I've just changed the file name) :
```
# proxmox-ccm.yaml
clusters:
  - url: https://cluster-api-1.exmple.com:8006/api2/json
    insecure: false
    token_id: "kubernetes@pve!ccm"
    token_secret: "secret"
    region: cluster-1
```

### Deploy by Helm the CCM ###
By deploying by the method 1 via kubectl didn't work for me. So, I've deployed by Helm with this command :
`helm upgrade -i --namespace=kube-system -f proxmox-ccm.yaml proxmox-cloud-controller-manager oci://ghcr.io/sergelogvinov/charts/proxmox-cloud-controller-manager`

### Create the CSI secret ###
Follow the doc (I've changed the file name and add/remove some options) :
```
# proxmox-csi.yaml
clusters:
  - url: https://cluster-api-1.exmple.com:8006/api2/json
    insecure: false
    token_id: "kubernetes-csi@pve!csi"
    token_secret: "secret"
    region: Region-1
node:
  nodeSelector:
  tolerations:
    - operator: Exists
nodeSelector:
  node-role.kubernetes.io/control-plane: ""
tolerations:
  - key: node-role.kubernetes.io/control-plane
    effect: NoSchedule
storageClass:
  - name: proxmox-data
    storage: local
    reclaimPolicy: Delete
    fstype: ext4
    cache: none
```

### Deploy by Helm the CCM ###
By deploying by the method 1 via kubectl didn't work for me. So, I've deployed by Helm with this command :
`helm upgrade -i --namespace=csi-proxmox -f proxmox-csi.yaml proxmox-csi-plugin oci://ghcr.io/sergelogvinov/charts/proxmox-csi-plugin`

### Test the set up ###
Wait for all the stuff to be created and then apply this test pod :
```
# test-pod-ephemeral.yaml
apiVersion: v1
kind: Pod
metadata:
  name: test
  namespace: default
spec:
  containers:
    - name: alpine
      image: alpine
      command: ["sleep","6000"]
      volumeMounts:
        - name: pvc
          mountPath: /mnt
  volumes:
    - name: pvc
      ephemeral:
        volumeClaimTemplate:
          metadata:
            labels:
              type: pvc-volume
          spec:
            accessModes: [ "ReadWriteOnce" ]
            storageClassName: proxmox-data
            resources:
              requests:
                storage: 1Gi
```

Please wait a little minute for the PV to be created and attached to the newly created test pod.

**Comment by sergelogvinov** [comment]
@LeoShivas Thanks, yep, `nodeSelector` is a bug.
I'll update the chart/result-deploy tomorrow...

**Comment by zimmertr** [comment]
Is CCM a dependency? This comment in the README led me to believe it was optional:

> Caution: set the labels topology.kubernetes.io/region and topology.kubernetes.io/zone are very important. **You can set it by kubectl or use [Proxmox CCM](https://github.com/sergelogvinov/proxmox-cloud-controller-manager)**. It uses the same Proxmox Cloud config. And it labels the node properly. I recommend using the CCM (Cloud Controller Manager).

Not because I don't want to use the CCM project, but because I'm just trying to learn one thing at a time 😄 

Otherwise, the instructions provided by @LeoShivas are identical to what I'm done already. 

<hr>

To @sergelogvinov's original feedback, I did add the additional `node.cloudprovider.kubernetes.io/platform: nocloud` labels. And that has helped me make progress. The DaemonSet is now creating `node` pods:

```bash
$> kubectl get pods -n csi-proxmox
NAME                                             READY   STATUS    RESTARTS   AGE
proxmox-csi-plugin-controller-6d5dd8856d-l96qj   5/5     Running   0          3m28s
proxmox-csi-plugin-node-ljxqb                    3/3     Running   0          3m28s
proxmox-csi-plugin-node-mhm6b                    3/3     Running   0          3m28s
proxmox-csi-plugin-node-pnfth                    3/3     Running   0          3m28s
```
And the `csinode` resources now have a non-null spec:
```bash
$> kubectl describe csinode k8s-node-1
Name:               k8s-node-1
Labels:             <none>
Annotations:        storage.alpha.kubernetes.io/migrated-plugins:
                      kubernetes.io/aws-ebs,kubernetes.io/azure-disk,kubernetes.io/azure-file,kubernetes.io/cinder,kubernetes.io/gce-pd,kubernetes.io/vsphere-vo...
CreationTimestamp:  Mon, 16 Oct 2023 15:26:07 -0700
Spec:
  Drivers:
    csi.proxmox.sinextra.dev:
      Node ID:  k8s-node-1
      Allocatables:
        Count:        16
      Topology Keys:  [topology.kubernetes.io/region topology.kubernetes.io/zone]
Events:               <none>


However, there was still no `csistoragecapacities` resources created:

```bash
$ > kubectl get csistoragecapacities.storage.k8s.io -A
No resources found
```

I interrogated all the logs again and grepped for `capacity` and found the following in the Controller's `csi-provisioner` container


> I1016 22:26:06.904186       1 capacity.go:373] Capacity Controller: storage class local was updated or added
I1016 22:26:06.904268       1 capacity.go:375] Capacity Controller: **ignoring storage class local because it uses immediate binding**

I I updated that line in the StorageClass and redeployed and voila! 

```bash
$> kubectl get csistoragecapacities.storage.k8s.io -A
NAMESPACE     NAME          CREATED AT
csi-proxmox   csisc-rjvr2   2023-10-16T22:43:02Z
```

I then deployed a test pod and got a volume and a running pod!

```bash
> kubectl get pvc,pv,pod
NAME                             STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS   AGE
persistentvolumeclaim/test-pvc   Bound    pvc-8cb4e355-8362-4359-a325-a3eb12e07767   1Gi        RWO            local          3m29s

NAME                                                        CAPACITY   ACCESS MODES   RECLAIM POLICY   STATUS   CLAIM                                      STORAGECLASS   REASON   AGE
persistentvolume/pvc-8cb4e355-8362-4359-a325-a3eb12e07767   1Gi        RWO            Delete           Bound    default/test-pvc                           local                   3m27s

NAME       READY   STATUS    RESTARTS   AGE
pod/test   1/1     Running   0          3m29s
```

And I can see the disk was created on Proxmox and added to the worker node running the pod!

![Screenshot 2023-10-16 at 3 48 45 PM](https://github.com/sergelogvinov/proxmox-csi-plugin/assets/4356289/8af3a3ed-12a8-4dc0-9ea2-8faa0c10a986)
![Screenshot 2023-10-16 at 3 48 24 PM](https://github.com/sergelogvinov/proxmox-csi-plugin/assets/4356289/ac545cf9-4b7c-479f-8f76-e48d48c63371)

This is very exciting! I can't wait to test it out with some real workloads!

**Comment by zimmertr** [comment]
I see [here](https://github.com/sergelogvinov/proxmox-csi-plugin/blob/main/docs/options.md#volumebindingmode) that `Immediate` is intended to be a supported  VolumeBindingMode. So perhaps this is a bug? Or maybe the `local` Storage type in Proxmox is incompatible with `Immediate`? 

I assume this will work fine with other Storage types like ZFS? Proxmox itself is probably managing all of that and Kubernetes and the CSI driver is otherwise agnostic?

**Comment by zimmertr** [comment]
A few followup questions now that I've had an opportunity to try this out. (It works great btw!) 

1. How will this work with multiple replicas? I'm assuming it would just dynamically provision multiple volumes and that `ReadWriteOnce` is the only supported AccessMode? Will `ReadOnlyMany` ever work?

2. How does the CSI driver react to pods moving around different virtual machines in a _Zone_? Does it detach the volume from a VM and attach it to the new VM before letting the pod start? Or is the pod intended to persist on a single node through an affinity or something after being provisioned?

3. Is there a known maximum of SCSI devices that can be attached to a VM through this provisioner? I tried Googling for it but I see varying results on the internet from 12 to 36 devices. I assume in very large clusters with very large nodes one could easily hit that limitation. 

4. Is static provisioning possible? I've manually created a ZFS volume called `FlashPool/vm-999-pvc-test` and seeded it with some data. Then I defined and applied a Pod, PV, and PVC like so. My goal in doing so, would be to give an existing set of data to a new pod. 

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: test
spec:
  containers:
    - name: alpine
      image: alpine
      command: ["sleep","6000"]
      volumeMounts:
        - name: test
          mountPath: /mnt/test
  volumes:
    - name: test
      persistentVolumeClaim:
        claimName: test
---
apiVersion: v1
kind: PersistentVolume
metadata:
  name: pvc-test
spec:
  accessModes:
    - ReadWriteOnce
  capacity:
    storage: 100Gi
  csi:
    driver: csi.proxmox.sinextra.dev
    fsType: ext4
    volumeAttributes:
      cache: writethrough
      ssd: "true"
      storage: FlashPool
      storage.kubernetes.io/csiProvisionerIdentity: 1697505561697-5065-csi.proxmox.sinextra.dev
    volumeHandle: sol-milkyway/earth/FlashPool/vm-999-pvc-test
  mountOptions:
    - noatime
  nodeAffinity:
    required:
      nodeSelectorTerms:
      - matchExpressions:
        - key: topology.kubernetes.io/region
          operator: In
          values:
            - sol-milkyway
        - key: topology.kubernetes.io/zone
          operator: In
          values:
            - earth
  storageClassName: proxmox-flashpool
---
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: test
spec:
  accessModes:
    - ReadWriteOnce
  resources:
    requests:
      storage: 100Gi
  storageClassName: proxmox-flashpool
  volumeName: pvc-test

```

When I apply it the pod gets stuck in pending. The PV and PVC are both bound successfully. And I can see in Proxmox an API call was made that keeps getting executed in a loop after it times out. It returned this output:

> update VM 4031: -scsi1 FlashPool:vm-9999-pvc-test,cache=writethrough,wwn=0x5056432d49443031,backup=0,iothread=1,ssd=1,discard=on
TASK ERROR: timeout: no zvol device link for 'vm-9999-pvc-test found after 300 sec found.

I looked through the logs and don't see any obvious errors. Just the GRPC calls to Proxmox to actually provision the disk. 

However, the dataset does exist:

```bash
$> zfs list -rt all | grep pvc-test
FlashPool/vm-9999-pvc-test               13.4G   600G     13.4G  /mnt/FlashPool/vm-9999-pvc-test
```

**Comment by sergelogvinov** [comment]
> I see [here](https://github.com/sergelogvinov/proxmox-csi-plugin/blob/main/docs/options.md#volumebindingmode) that `Immediate` is intended to be a supported VolumeBindingMode. So perhaps this is a bug? Or maybe the `local` Storage type in Proxmox is incompatible with `Immediate`?

Never tried the Immediate disks.. ) the error goes from kubernetes side. Probably we need to check official documentations.

**Comment by sergelogvinov** [comment]
> A few followup questions now that I've had an opportunity to try this out. (It works great btw!)
> 
> 1. How will this work with multiple replicas? I'm assuming it would just dynamically provision multiple volumes and that `ReadWriteOnce` is the only supported AccessMode? Will `ReadOnlyMany` ever work?

Volume can exists (attached) only in one VM. Inside one VM, many pods can have RW-access to one PVC. 

> 2. How does the CSI driver react to pods moving around different virtual machines in a _Zone_? Does it detach the volume from a VM and attach it to the new VM before letting the pod start? Or is the pod intended to persist on a single node through an affinity or something after being provisioned?

It dose scheduler:  stop pod -> unmount pvc on VM -> detach PVC (volume) from node -> scheduler choses the new node -> csi attaches the pvc to new node -> mount process inside the VM -> run the pod.

PVC cannot migrate between proxmox nodes (migration does not support yet)  - thats the reason why ZONE - is proxmox node.

> 3. Is there a known maximum of SCSI devices that can be attached to a VM through this provisioner? I tried Googling for it but I see varying results on the internet from 12 to 36 devices. I assume in very large clusters with very large nodes one could easily hit that limitation.

Two limitation exists: proxmox side, csi side. https://pve.proxmox.com/wiki/Manual:_qm.conf - `Use volume as SCSI hard disk or CD-ROM (n is 0 to 30).` CSI has limit  - 16. we can change it to 28. 
I think we need to save a few spots to manual attaches disk...

So currently it is 16, but there is not technical limitation to increase it.

> 4. Is static provisioning possible? I've manually created a ZFS volume called `FlashPool/vm-999-pvc-test` and seeded it with some data. Then I defined and applied a Pod, PV, and PVC like so. My goal in doing so, would be to give an existing set of data to a new pod.

Try to create the PVC as usual, and check/compare the result of `volume_id`, probably this is a problem.
The error `timeout: no zvol device link` goes from Proxmox side.
Or attache Volume manually (proxmox gui) and compere the VM config result `scsiX: ...`

**Comment by zimmertr** [comment]
So I'm able to use the CSI driver to dynamically provision a volume via my ZFS storage. Then I can delete the pod, pv, pvc but retain the ZFS dataset. Then I can successfully create a new pod, pv, pvc referencing the dynamically created ZFS dataset and the pod will come up just fine.

However, when I try and create a pod, pv, pvc referencing a ZFS dataset I created by hand, it never comes up. Even if the Pod, PV, and PVC YAML is exactly the same (with the exception of the dataset name)

Proxmox just loops returning this error:

> update VM 4031: -scsi1 FlashPool:vm-9999-pvc-static-pod-test,cache=writethrough,wwn=0x5056432d49443031,backup=0,iothread=1,ssd=1,discard=on
TASK ERROR: timeout: no zvol device link for 'vm-9999-pvc-static-pod-test' found after 300 sec found. 

Is there something unique about the way you're creating datasets with the CSI driver? Is it different than a simple:

```bash
$> zfs create FlashPool/vm-9999-pvc-static-pod-test
```

**Comment by zimmertr** [comment]
Okay, I've figured it out. Looks like it must be a ZFS volume, not a ZFS dataset. I discovered this by running `zfs get all` against a dynamically created volume and the statically created dataset. Then comparing the properties. 

tl;dr: Create them like this instead: `zfs create -V 10G FlashPool/vm-9999-pvc-static-pod-test` 

I think this can be closed now. Thank you again for all of your help! If you enable sponsorship for this repo I will start sponsoring your work.
